### ISSUE TITLE
Endian Problems in snappy on big endian systems

### ISSUE THREAD

**Issue by rthalley** [edenhill/librdkafka#1156]
Description
===========

There are multiple problems with snappy on big endian systems.  The first problem is that this code in snappy-compat.h doesn't work

```#if __BYTE_ORDER == __LITTLE_ENDIAN
#define __LITTLE_ENDIAN__ 1
#endif
```

Neither __BYTE_ORDER nor __LITTLE_ENDIAN are defined, so it ends up comparing 0 == 0, which is true, and so it sets __LITTLE_ENDIAN__ when the system isn't little endian.  (This surprised me, but it's correct ISO C behavior, I checked!)  The patch below fixes that issue, and I've got all relevant patches from your "align branch" applied too, yet still there are problems writing though reading seems to work.  I do not currently know why writing fails; all I know is the broker complains with a " Broker: Invalid message" error.  Setting "compression.codec" to "none" allows writes to work, so I assume the problem is snappy related.  A colleague and I looked for other obvious endian problems in snappy but we didn't notice any.

I looked to see if there are any upstream fixes in snappy but didn't see any.  Upstream doesn't compile on SPARC at all, nor does it have a test suite, so I haven't tried further upstream debugging.

How to reproduce
================

Compile with the patch below on solaris sparc.   Send a message to a broker with snappy compression enabled; the broker will reject it.

Checklist
=========
Please provide the following information:

 - [ ] librdkafka version (release number or git tag): v0.9.4
 - [ ] Apache Kafka version: 0.10.2
 - [ ] librdkafka client configuration: compression.codec = snappy
 - [ ] Operating system: Solaris 11 SPARC
 - [ ] Using the legacy Consumer N/A
 - [ ] Using the high-level KafkaConsumer N/A
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka N/A
 - [ ] Provide broker log excerpts N/A
 - [ ] Critical issue yes

```
diff -u -r old/src/rdendian.h new/src/rdendian.h
--- old/src/rdendian.h  Fri Apr  8 08:01:44 2016
+++ new/src/rdendian.h  Tue Apr  4 07:15:38 2017
@@ -63,14 +63,30 @@
 #elif defined sun
   #include <sys/byteorder.h>
   #include <sys/isa_defs.h>
+#define __LITTLE_ENDIAN 1234
+#define __BIG_ENDIAN 4321
 #ifdef _BIG_ENDIAN
+#define __BYTE_ORDER __BIG_ENDIAN
 #define be64toh(x) (x)
 #define be32toh(x) (x)
 #define be16toh(x) (x)
+#define le16toh(x) BSWAP_16(x)
+#define le32toh(x) BSWAP_32(x)
+#define le64toh(x) BSWAP_64(x)
+#define htole16(x) BSWAP_16(x)
+#define htole32(x) BSWAP_32(x)
+#define htole64(x) BSWAP_64(x)
 # else
+#define __BYTE_ORDER __LITTLE_ENDIAN
 #define be64toh(x) BSWAP_64(x)
 #define be32toh(x) ntohl(x)
 #define be16toh(x) ntohs(x)
+#define le16toh(x) (x)
+#define le32toh(x) (x)
+#define le64toh(x) (x)
+#define htole16(x) (x)
+#define htole32(x) (x)
+#define htole64(x) (x)
 #endif /* sun */
 
 #elif defined __APPLE__
```

**Comment by edenhill** [comment]
Ouch, that's no good.
The snappy code is indeed quite hard to follow and I don't have access to a SunOS box, or any big endian that I know of. I guess one could set up a big-endian arm build and use qemu, but that's a bit of work.

Would using lz4 be an alternative for you?

Another alternative would be to add support for libsnappy.

**Comment by rthalley** [comment]
We pretty much need snappy, lz4 was too slow as I recall.  This isn't a showstopper for me at the moment, but I wanted to open a ticket so the problem was known in case anyone else hit it.  Or in case other eyes looked at snappy and said "oh yeah, that's it" :)  If I get some spare time in the future, I may try to figure out more about what is bad.

**Comment by rthalley** [comment]
We figured this one out.

The crucial changes are in the following patch.  The two key things are 1) getting __LITTLE_ENDIAN, __BIG_ENDIAN, and __BYTE_ORDER defined properly, and 2) the (uint16_t) part of the +#define htole16(x) ((uint16_t)BSWAP_16(x)) line.  The cast is essential as otherwise in this macro in snappy_compat.h

```#define put_unaligned_memcpy(v,x) ({ \
		typeof((v)) _v = (v); \
		memcpy((x), &_v, sizeof(*(x))); })
```

the value of 'v' has typeof int, not uint16_t, and it takes the 32-bit extended value and copies the first two bytes, which being big endian, are zero and not the right swapped value.

```diff -u -r old/src/rdendian.h new/src/rdendian.h
--- old/src/rdendian.h	Fri Apr  8 08:01:44 2016
+++ new/src/rdendian.h	Tue Apr  4 07:15:38 2017
@@ -63,14 +63,30 @@
 #elif defined sun
   #include <sys/byteorder.h>
   #include <sys/isa_defs.h>
+#define __LITTLE_ENDIAN 1234
+#define __BIG_ENDIAN 4321
 #ifdef _BIG_ENDIAN
+#define __BYTE_ORDER __BIG_ENDIAN
 #define be64toh(x) (x)
 #define be32toh(x) (x)
 #define be16toh(x) (x)
+#define le16toh(x) ((uint16_t)BSWAP_16(x))
+#define le32toh(x) BSWAP_32(x)
+#define le64toh(x) BSWAP_64(x)
+#define htole16(x) ((uint16_t)BSWAP_16(x))
+#define htole32(x) BSWAP_32(x)
+#define htole64(x) BSWAP_64(x)
 # else
+#define __BYTE_ORDER __LITTLE_ENDIAN
 #define be64toh(x) BSWAP_64(x)
 #define be32toh(x) ntohl(x)
 #define be16toh(x) ntohs(x)
+#define le16toh(x) (x)
+#define le32toh(x) (x)
+#define le64toh(x) (x)
+#define htole16(x) (x)
+#define htole32(x) (x)
+#define htole64(x) (x)
 #endif /* sun */
 
 #elif defined __APPLE__
```

I also had this patch taken from your master branch

```diff -u -r old/src/rdkafka_broker.c new/src/rdkafka_broker.c
--- old/src/rdkafka_broker.c    2017-03-03 06:58:24.000000000 -0800
+++ new/src/rdkafka_broker.c    2017-04-03 13:39:28.000000000 -0700
@@ -3394,11 +3394,13 @@
 
                while (of + 4 <= (ssize_t)inlen) {
                        /* compressed length */
-                       uint32_t clen = be32toh(*(uint32_t *)(inbuf+of));
+                       uint32_t clen;
                        /* uncompressed length */
                        size_t ulen;
                        int r;
 
+                       memcpy(&clen, inbuf+of, 4);
+                       clen = be32toh(clen);
                        of += 4;
 
                        if (unlikely(clen > inlen - of)) {
```

I'm not sure that latter one is essential, but I mention it since we had it when testing.

With those patches, we now have working send and receive with snappy on Solaris 11 SPARC.

**Comment by rthalley** [comment]
oops, shouldn't have closed it; that's for you :)

**Comment by edenhill** [comment]
@rthalley Do you think you could file a PR for this?

**Comment by rthalley** [comment]
done
